Rt1170 Dual Core and OpenAMP support

danieldegrasse opened the pull request:
Enable dual core and shared memory support for the RT1170 EVK.

This PR enables the following features for the RT1170:
- Dual core support (primary core can kick off secondary core without debugger)
- Shared memory support using `OCRAM2`
- Messaging unit driver to enable communication between cores


~Mulitcore images are created by converting the secondary image binary into a elf file with one section, and linking that into the primary image. The primary core will load the secondary image from flash into RAM, then boot the secondary core from that RAM.~
Mulicore images are now supported using Sysbuild

danieldegrasse committed:
boards: mimxrt1170_evk: Refactor flexspi partition definition

Flexspi partition definition is shared between both cm4 and cm7 core.
Refactor it to be in shared dtsi file. This will ensure that both cores
share the same partition information, for dual core operation.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt1170_evk: Enable LED1 node

LED1 node is enabled, to provide a secondary led for the CM4 core.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
dts: nxp_rt11xx: Refactored RT11xx CM7 and CM4 DTS

RT1170 and RT1160 CM7 and CM4 cores have the same set of differences.
Merge the DTS files for both CM4 and CM7 cores, to create generic
rt11xx_cm4 and rt11xx_cm7 files.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: nxp_imx: Add code to wait for second core boot in RT11xx

Use the messaging unit to ensure that the RT11xx dual core mode will
wait for the second core to boot successfully during early init

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
dts: nxp_imx: Add zephyr,memory-region attribute to memory regions

Add zephyr,memory-region compatible and attribute to SOC memory regions,
so that sections will be generated and MPU attributes can be applied.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
samples: ipc: openamp: Enable openamp sample for iMX.RT1170 EVK

Enable the openamp sample for RT1170EVK. CM4 core has a custom DTS
overlay to use LPUART2 for console information, and use a secondary GPT
timer for the system tick.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt1170_evk: Update RT1170 documentation for dual core support

Update RT1170 documentation for dual core support

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

carlocaione commented:
Just for the record: a lot of effort was lately done to move from IPM to MBOX, and from OpenAMP samples to IPC service.

danieldegrasse commented:
@carlocaione Thanks for the heads up- I was unaware. Should I target enabling the `static_vrings_mi` sample on this board when I create an MBOX driver for it?

carlocaione commented:
> @carlocaione Thanks for the heads up- I was unaware. Should I target enabling the `static_vrings_mi` sample on this board when I create an MBOX driver for it?

If you are willing to write a new MBOX drive I suggest you to enable the `samples/subsys/ipc/ipc_service/` sample. The `static_vrings_mi` is going to be deleted (in my TODO list :cry:  )

dleach02 commented:
@carlocaione is the project going to remove the IPM drivers? To be honest, I'm not up on the differences between IPM and MBOX but it seems like if the project has IPM support and MBOX support, it can be left up to the application to decide which way to enable this?

carlocaione commented:
> @carlocaione is the project going to remove the IPM drivers?

No, this was discussed in the past and it was decided to keep both.

> To be honest, I'm not up on the differences between IPM and MBOX

MBOX supports multiple channels (for example several devices using the MBOX instance at the same time), IPM does not.

danieldegrasse commented:
@carlocaione So to clarify, mbox drivers may use the same peripherals as ipm drivers, but they expose channels to allow multiple users of the IPM peripheral? In other words, is the mbox api essentially a new api to interact with the same peripheral as an ipm driver would use?

carlocaione commented:
> So to clarify, mbox drivers may use the same peripherals as ipm drivers, but they expose channels to allow multiple users of the IPM peripheral? 

Correct.

> In other words, is the mbox api essentially a new api to interact with the same peripheral as an ipm driver would use?

Yes. The MBOX API was introduced because sometimes you want to share the peripheral between multiple users.  One notable example is when you have multiple OpenAMP instances that want to use the same mailbox peripheral for signalling. In that case you cannot use IPM but you have to use MBOX. For reference see:

https://github.com/zephyrproject-rtos/zephyr/blob/c9a3dae6519fc788d33233fc1c4d1364ba129034/samples/subsys/ipc/ipc_service/static_vrings/boards/nrf5340dk_nrf5340_cpuapp.overlay#L24-L56

In this case both (`ipc0` and `ipc1`) OpenAMP instances are using the same mbox peripheral (but different channels) at the same time.

SebastianBoe commented on soc/arm/nxp_imx/rt/mpu_regions.c:
```diff
@@ -0,0 +1,30 @@
+ */
+
+#include <devicetree.h>
+#include "../../common/cortex_m/arm_mpu_mem_cfg.h"
```
I see you are not the first to need this header file.

Shall we do zephyr_include_directories(.) from soc/arm/common/cortex_m/CMakeLists.txt ?

Or perhaps even better move the header to

zephyr/include/arch/arm/aarch32/mpu/

```
sebo@sebo:~/ncs/bootloader/mcuboot$ ack arm_mpu_mem_cfg.h ~/ncs/
/home/sebo/ncs/zephyr/soc/arm/common/cortex_m/arm_mpu_regions.c
10:#include "arm_mpu_mem_cfg.h"

/home/sebo/ncs/zephyr/soc/arm/nuvoton_npcx/npcx7/mpu_regions.c
7:#include "../../common/cortex_m/arm_mpu_mem_cfg.h"

/home/sebo/ncs/zephyr/soc/arm/st_stm32/stm32h7/mpu_regions.c
8:#include "../../common/cortex_m/arm_mpu_mem_cfg.h"

```

carlocaione replied:
Also shameless plug for #41223 about this

danieldegrasse replied:
@SebastianBoe I agree that we should take one of those approaches with the file, but should it be something addressed by this PR, or a separate PR specifically calling out the issue and fixing it?

SebastianBoe replied:
A separate PR would be best.

tejlmand commented on soc/arm/nxp_imx/rt/CMakeLists.txt:
```diff
@@ -30,3 +30,37 @@ zephyr_linker_section_configure(
+	# one is linked
+    set_target_properties(remote_core_image PROPERTIES IMPORTED_LOCATION
+		${REMOTE_ZEPHYR_DIR}/${KERNEL_BIN_NAME})
+	target_link_libraries(app PUBLIC remote_core_image)
```
This looks like an artificial dependency.

`app` is not really depending upon `remote_core_image` as far as I can tell.

danieldegrasse replied:
It may be the wrong thing to depend on. The goal here is to force the primary image to depend on the remote image. Is there another target I could depend on besides `app` to ensure that dependency is enforced?

The reason for that requirement is that the final output dual core image is an image intended to run on the primary core, with the second core's image linked in as a section. This image is read from flash into RAM by the primary core, which then boots the secondary core from RAM.

tejlmand reviewed (changes requested):
This looks like yet another multiimage approach.

Tagged for devreview.

Please take a look here #40555 and provide feedback so that we can solve the recurring multi image problem in a generic way suitable for all use-cases and not invent individual solutions everywhere.

SebastianBoe commented on samples/subsys/ipc/openamp/CMakeLists.txt:
```diff
@@ -16,6 +16,8 @@ elseif("${BOARD}" STREQUAL "mps2_an521")
 elseif("${BOARD}" STREQUAL "v2m_musca_b1")
   set(BOARD_REMOTE "v2m_musca_b1_ns")
+elseif("${BOARD}" STREQUAL "mimxrt1170_evk_cm7")
+  set(BOARD_REMOTE "mimxrt1170_evk_cm4")
```
@tejlmand : AFAICT this PR just adds support for another platform to do multi-image in a sample that already does multi-image.

So it is following prior-art and should not given -1 for doing yet-another-multi-image approach.

SebastianBoe replied:
reading this CMakeLists.txt file in it's entirety makes this clearer.

danieldegrasse replied:
@SebastianBoe that's correct, this PR is another multi-image approach. I do agree that we need a generic multi image approach in Zephyr, however.

tejlmand replied:
> So it is following prior-art and should not given -1 for doing yet-another-multi-image approach.

Except for the following two things I spotted immediately.
There might be more if taking a closer look:
- The SoC code itself relies on the `REMOTE_ZEPHYR_DIR` when `DUAL_CORE_MCUX` is enabled and then does a conversion and embed the second image into the Zephyr image.
  No other SoC code does this directly, as far as I can see.
  This is a new approach
- The PR introduces `DUAL_CORE_MCUX` Kconfig setting to control when a second image should be included.
   A very generic name, yet located specific in a Kconfig.soc:
   https://github.com/zephyrproject-rtos/zephyr/blob/c10d9c4d1e39c7e533e7f48c484d7a731c26d2ee/soc/arm/nxp_imx/rt/Kconfig.soc#L656
  This is a new strategy that I would like to be discussed, especially as we have #40555 coming.
    So a new setting that might be soon deprecated is not nice.

SebastianBoe replied:
> embed the second image into the Zephyr image.

I could have sworn this is what the open-amp sample has always been doing. They don't do hex/bin merging.
But perhaps this has changed in the last couple of years.

SebastianBoe replied:
If this is not just following prior-art then the author should at least state why it is necessary to do something different.

tejlmand replied:
> > embed the second image into the Zephyr image.
> 
> I could have sworn this is what the open-amp sample has always been doing. They don't do hex/bin merging. But perhaps this has changed in the last couple of years.

This is the mess I would like to get cleaned up by having a uniform multi-image approach.
Unified in the sense it is not tied into single image Zephyr build.
Different multi-core SoC of course may have different ways they need to flash / embed multiple images.

So currently for `openamp` we have:
- lpcxpresso54114_m4 and lpcxpresso55s69_cpu0 which uses `generate_inc_file_for_gen_target()` and has a dedicated `SECOND_CORE_MCUX` setting
- mps2_an521 and v2m_musca_b1 which don't embed the second image, and don't have special setting.
   mps2_an521 even with the twist that if you build for cpu1 and not `openamp` then you always get a dummy image build for cpu0 that you cannot change. Not setting to disable and no direct Kconfig options for that image.

and now this PR which embeds the second image in another way not covered by the above methods but takes the bin file directly and embed as a section.

I prefer at this point we get a higher level handling of multi-image and thus spend time on such solution instead of yet another small variant for another SoC / board under `subsys/ipc`.

SebastianBoe commented on soc/arm/nxp_imx/rt/CMakeLists.txt:
```diff
@@ -30,3 +30,37 @@ zephyr_linker_section_configure(
+	# a dependency that the secondary code image be built before the primary
+	# one is linked
+    set_target_properties(remote_core_image PROPERTIES IMPORTED_LOCATION
+		${REMOTE_ZEPHYR_DIR}/${KERNEL_BIN_NAME})
```
See https://github.com/zephyrproject-rtos/zephyr/pull/41579/files#r779630103

danieldegrasse replied:
@tejlmand there is no current way to ensure `REMOTE_ZEPHYR_DIR` is set- but I think a build failure is the right way to approach the issue. I can add a check in the CMake file to provide a meaningful error to the user, but in the end selecting `CONFIG_DUAL_CORE_MCUX` tells the build system that there will be a second image present for the secondary core, so `REMOTE_ZEPHYR_DIR` must be set for the build to proceed.

danieldegrasse commented:
@tejlmand I've removed new KConfig symbols specific to this dual core implementation, and the soc level code now only builds for dual core if `CONFIG_OPENAMP` is selected.

Laczen commented:
@danieldegrasse, maybe a stupid question. Why is the image for the second core added to the flash of the primary ? Is it not possible to dedicate a piece of flash for the second core image ?

dleach02 commented:
> @tejlmand I've removed new KConfig symbols specific to this dual core implementation, and the soc level code now only builds for dual core if `CONFIG_OPENAMP` is selected.

I'm not sure `CONFIG_OPENAMP` is the right keyword to trigger on. We can certainly have a single embedded core on a larger system that enables openamp to talk to the larger (maybe Linux based) system. It feels like we want or need a config that is specific to multi-image for multi-CPU systems and not necessarily tie it to whether or not OpenAMP is used.

danieldegrasse commented:
@Laczen The requirement to run the second core from RAM means that the second image must be linked with the first. This is because the primary core image must know the size of the second core image in order to copy it from flash to ram.

To be clear, it is possible to boot the secondary core directly from flash, but this is not preferred because when both the M7 and M4 cores share access to the flash it will result in delayed instruction fetches and slower execution. Moreover, if both cores share access to the flash, there would need to be additional application code to handle flash writes, since one core writing to the flash while the other reads would create a RWW hazard

tejlmand commented:
@danieldegrasse based on the talks at dev-review meeting I have created a small PR which allows to generate a header file containing the information you said was needed to know for the first core.

Please take a look here: #41685.

This generates a header with the information you should need in the form:
```
define SEGMENT_NUM 4

#define SEGMENT_LMA_ADDRESS_0 0x20205be0
#define SEGMENT_VMA_ADDRESS_0 0x20205be0
#define SEGMENT_SIZE_0 0x8

#define SEGMENT_LMA_ADDRESS_1 0x20200000
#define SEGMENT_VMA_ADDRESS_1 0x20200000
#define SEGMENT_SIZE_2 0x6f3c

.....
```
[UTIL_LISTIFY](https://docs.zephyrproject.org/latest/reference/util/index.html#c.UTIL_LISTIFY) is useful to create an array of this data.

Please provide feedback as to how this would fit your needs.
If it does, I think #41685 will be useful together with #40555.

tejlmand commented:
> > It feels like we want or need a config that is specific to multi-image for multi-CPU systems and not necessarily tie it to whether or not OpenAMP is used.
>
> @tejlmand, thoughts?

Correct, we should get a lot of this knowledge out of the single image Zephyr build and then have a higher level doing the multi image work, which is what i'm working on here #40555.
This is part of higher level work here: https://github.com/zephyrproject-rtos/zephyr/issues/41640 which tries to track the various requirements needed.

Until that work is complete, we should try to simply contain current multi-image solutions inside `samples/subsys/ipc` and not have it spread to samples in general.
Only reason to still allow `subsys/ipc` to do custom multi-image is because they have been allowed in past where we had no general solution available.

danieldegrasse commented:
@tejlmand I've reworked this PR based off your work in #41685. The SoC level code still utilizes `CONFIG_OPENAMP` to determine if the M4 core should be booted from the M7, but I moved cmake changes into the openamp sample where possible. Please let me know your thoughts on this approach.

dleach02 commented:
Daniel, the HAL PR was merged so you can update the west.yaml.

danieldegrasse commented:
@dleach02 I've updated the PR to use the messaging unit driver introduced with #42320

dleach02 commented:
HAL PR was merged so this needs a change to the west.yaml

danieldegrasse commented:
@dleach02 rebased to include latest `west.yml`

dleach02 commented:
@tejlmand, @carlocaione can you revisit

danieldegrasse commented on soc/arm/nxp_imx/rt/CMakeLists.txt:
```diff
@@ -11,7 +11,10 @@ zephyr_linker_sources_ifdef(CONFIG_NXP_IMX_RT_BOOT_HEADER
-zephyr_sources_ifdef(CONFIG_OCRAM_NOCACHE mpu_regions.c)
+# MPU regions definition is required if openamp support is enabled, or for nocache region
+if(CONFIG_OCRAM_NOCACHE OR CONFIG_OPENAMP)
+zephyr_sources(mpu_regions.c)
```
I updated the PR to use that. Note- the MPU region attribute for the shared memory needs to be `REGION_IO_ATTR` instead of `REGION_RAM_NOCACHE_ATTR`, so I updated the DT binding for `zephyr,memory-region` to allow for "IO" to be one of the possible `zephyr,memory-region-mpu` settings.

carlocaione reviewed (dismissed):
I'm not an expert on this platform but my comments were addressed.

danieldegrasse commented:
@tejlmand Can you revisit this PR? Given the fact that separate images are being built, and there is not any custom solution for `west flash` used by this PR, I think that transitioning this multi image support from its current state to something compatible with sysbuild will be relatively simple when sysbuild merges. In the meantime, I'd really like to see this PR go in: It will provide a sample for customers showing how to boot the M4 core present on the RT1170, which they can extend for their specific use case.

mbolivar-nordic commented:
@danieldegrasse unfortunately @tejlmand is out on parental leave and won't be back in time for code freeze. If you want this for v3.1, I suggest you apply for an exception from the TSC to see if dismissing his review and merging for v3.1 after code freeze is acceptable.

danieldegrasse commented:
@mbolivar-nordic I'll discuss with @dleach02 to see if that is something we'd like to do, thanks for the heads up

tyalie commented:
Are there any news on this now that #40555 has been merged?

danieldegrasse commented:
This PR is dependent on #48709, which itself depends on #49552. I've kept this PR open in the meantime because I would like to have it available as a demonstration of how to support multicore operation on the RT11xx series. However, it may be some time before this is able to merge.

danieldegrasse committed:
samples: ipc: openamp: Enable openamp sample for iMX.RT1160 EVK

Enable the openamp sample for RT1160EVK. CM4 core has a custom DTS
overlay to use LPUART2 for console information, and use a secondary GPT
timer for the system tick.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: rt1xxx: allow linking code to OCRAM region

Allow linking code into OCRAM region when building for RT1xxx SOCs. This
can be used on the RT11xx dual core SOCs as a shared memory region, when
the M7 core needs to load code into a region accessible to the M4 core.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: rt11xx: Enabled multicore support with second image

RT11xx features CM4 core, which must be booted from CM7 core. Add
support for loading an image for the CM4 to RAM, and booting the CM4 core
from this image. Each image is built independently using sysbuild, and the
M4 image build produces built collateral with load address information the
M7 image can use to load it to RAM

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt1170_evk: add support for loading M4 image from OCRAM

Add support for loading M4 image from OCRAM when running with dual core
operation. The M7 core will copy the M4 image from flash into OCRAM, and
the M4 core will execute it there.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: ipc: Enable messaging unit driver for iMX.RT multicore SOCs

Enable MU messaging unit driver for RT11xx socs.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt1170_evk: enable messaging unit

Enable messaging unit support for RT1170 EVK

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt1160_evk: Refactor flexspi partition definition

Flexspi partition definition is shared between both cm4 and cm7 core.
Refactor it to be in shared dtsi file. This will ensure that both cores
share the same partition information, for dual core operation.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt1160_evk: add support for loading M4 image from OCRAM

Add support for loading M4 image from OCRAM when running with dual core
operation. The M7 core will copy the M4 image from flash into OCRAM, and
the M4 core will execute it there.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt1160_evk: enable messaging unit

Enable messaging unit support for RT1160 EVK

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt1160_evk: Update RT1160 documentation for dual core support

Update rt1160 documentation for dual core support

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

tejlmand reviewed (approved):
lgtm
